storage/storage_dest.go

```diff
@@ -317,122 +325,141 @@ func (s *storageImageDestination) TryReusingBlobWithOptions(ctx context.Context,
+				layers, err := s.imageRef.transport.store.LayersByUncompressedDigest(uncompressedDigest)
+				if err != nil && !errors.Is(err, storage.ErrLayerUnknown) {
+					return false, private.ReusedBlob{}, fmt.Errorf(`looking for layers with digest %q: %w`, uncompressedDigest, err)
+				}
+				if len(layers) > 0 {
+					if size != -1 {
+						s.blobDiffIDs[digestBlob] = layers[0].UncompressedDigest
+						return true, private.ReusedBlob{
+							Digest: digestBlob,
+							Size:   size,
+						}, nil
+					}
+					if !options.CanSubstitute {
+						return false, private.ReusedBlob{}, fmt.Errorf("Internal error: options.CanSubstitute was expected to be true for blob with digest %s", digestBlob)
+					}
+					s.blobDiffIDs[uncompressedDigest] = layers[0].UncompressedDigest
 					return true, private.ReusedBlob{
-						Digest: digest,
-						Size:   size,
+						Digest: uncompressedDigest,
+						Size:   layers[0].UncompressedSize,
 					}, nil
 				}
-				if !options.CanSubstitute {
-					return false, private.ReusedBlob{}, fmt.Errorf("Internal error: options.CanSubstitute was expected to be true for blob with digest %s", digest)
-				}
-				s.uncompressedOrTocDigest[uncompressedDigest] = layers[0].UncompressedDigest
-				return true, private.ReusedBlob{
-					Digest: uncompressedDigest,
-					Size:   layers[0].UncompressedSize,
-				}, nil
 			}
 		}
 	}
 
-	tocDigest := digest
-	if options.TOCDigest != nil {
-		tocDigest = *options.TOCDigest
-	}
+	if options.TOCDigest != "" {
+		// we are already holding s.lock
+		diffOutput, ok := s.diffOutputs[digestBlob]
+		if ok {
+			// Save this for completeness.
+			s.tocDigests[options.TOCDigest] = options.TOCDigest
+			return true, private.ReusedBlob{
+				TOCDigest: diffOutput.TOCDigest,
+				Size:      diffOutput.Size,
+			}, nil
+		}
 
-	// Check if we have a chunked layer in storage with the same TOC digest.
-	layers, err = s.imageRef.transport.store.LayersByTOCDigest(tocDigest)
-	if err != nil && !errors.Is(err, storage.ErrLayerUnknown) {
-		return false, private.ReusedBlob{}, fmt.Errorf(`looking for layers with TOC digest %q: %w`, tocDigest, err)
-	}
-	if len(layers) > 0 {
-		// Save this for completeness.
-		s.uncompressedOrTocDigest[digest] = layers[0].TOCDigest
-		return true, private.ReusedBlob{
-			Digest: layers[0].TOCDigest,
-			Size:   layers[0].UncompressedSize,
-		}, nil
+		// Check if we have a chunked layer in storage with the same TOC digest.
+		layers, err := s.imageRef.transport.store.LayersByTOCDigest(options.TOCDigest)
+
+		if err != nil && !errors.Is(err, storage.ErrLayerUnknown) {
+			return false, private.ReusedBlob{}, fmt.Errorf(`looking for layers with TOC digest %q: %w`, options.TOCDigest, err)
+		}
+		if len(layers) > 0 {
+			// Save this for completeness.
+			s.tocDigests[options.TOCDigest] = layers[0].TOCDigest
+			return true, private.ReusedBlob{
+				TOCDigest: layers[0].TOCDigest,
+				Size:      layers[0].UncompressedSize,
```

Comment:
Is this value… known? AFAICS right now it comes from `DriverWithDifferOutput`… which never sets the field.

Yet, it is a mandatory manifest field, so we must provide the “correct” value.

That, again, suggests preserving the original `Digest`+`Size` here.

Reply:
That value would, conceptually, be written into the manifest stored via `PutManifest`. And, see elsewhere, we primarily want to preserve the manifest unmodified, along with original signatures.

… I think `layerDigestsDiffer` is involved and it means some of the updates are (correctly? incorrectly?) just not done. I’ll might need to revisit what actually happens.

Reply:
Locally, if this were the _only_ change on pull, we wouldn’t update layer info in the manifest (we update the manifest either if `LayerInfosForCopy` returned something, or if at least one digest changes).

But it’s possible for some other layer entry to change the digest, and then the values returned here would be used, even if _this_ layer didn’t return a changed digest. Besides, returning inconsistent metadata is just unclean.

#2294 should fix this.